src/commands/command.ts

```diff
@@ -0,0 +1,53 @@
+import { Message } from 'discord.js';
+import { PermissionsLevel } from '@/services/permissions.service';
+
+export type TriggerCriteria = string | RegExp | ((message: Message, keyword: string, args: string[]) => string | null);
+
+export interface ConcreteTrigger {
+    command: Command,
+    criteria: TriggerCriteria,
+    type: 'string' | 'function' | 'regexp',
+    activations: { text: string, index: number }[]
+}
+
+export abstract class Command {
+    abstract name: string;
+    abstract description: string;
+    abstract triggerCriteria: TriggerCriteria[];
+    requiredPerms: PermissionsLevel = PermissionsLevel.USER;
+
+    abstract validate(message: Message, trigger: ConcreteTrigger, args: string[]): boolean;
+    abstract execute(message: Message, trigger: ConcreteTrigger, args: string[]): void;
+
+    checkTriggers(message: Message, keyword: string, args: string[]): ConcreteTrigger | null {
+        const trigger: Partial<ConcreteTrigger> = {
+            command: this
+        };
+
+        for (const criteria of this.triggerCriteria) {
+            if (typeof criteria === 'function') {
+                const res: string | null = criteria(message, keyword, args);
+                if (!res || res.length < 1) continue;
+                trigger.type = 'function';
+                trigger.activations = [{ text: res, index: -1 }];
+            }
+            if (criteria instanceof RegExp) {
+                const res: RegExpMatchArray[] = Array.from(message.content.matchAll(criteria));
+                if (res.length < 1) continue;
+                trigger.type = 'regexp';
+                trigger.activations = res.map((m: RegExpMatchArray) => ({
+                    text: m[0] ?? message.content,
+                    index: m.index ?? -1
+                }));
+            }
+            if (typeof criteria === 'string') {
+                if (criteria !== keyword) continue;
+                trigger.type = 'string';
+                trigger.activations = [{text: keyword, index: 0}];
+            }
+            trigger.criteria = criteria;
+            return <ConcreteTrigger>trigger;
+        }
+        return null;
+    }
```

Comment:
Really smart to put the `checkTriggers` into Command instead of into the Message Event! The `Partial` Syntax is new to me, so I'll have to look through what it does!  
As for the `activations` - which you've added in here, I'll assume it saves the method of activation / precise wording of the command, but didn't you mean to *`push({text: res, index: -1})`* instead of *assigning* it to activations?  
Also, another note for me - the `??` syntax is probably a mix out of *if not null* and *ternary op.*.
I don't quite understand why you'd save activations, and type - but maybe you plan on using it.  
I suggest on expanding the class with an `addTrigger` method - but that's something for later, to add triggers udring Runtime.  
Great thing that you already did Perms, that's what I would've done next!

Reply:
Meant to assign it. In this case, I'm counting it as if there's only one trigger string, so assignment is correct in all places, especially since I haven't initialized it to an empty array anywhere. You'll notice that in the second case, with the Regular expression, I'm actually assigning it to an array (the result of the `.map()`) --- this is honestly the only reason why I'm making this an array, but that way if multiple parts of the string match the regexp, that's noted.

See above for `??`. `??` is the same as `||` except where  `a || b` results in `b` for any falsy `a`,  `a ?? b` results in `b` only if `a` is `null` or `undefined`. It's a TypeScript 3.7/ES2020 thing.

See above for activations/type.

addTrigger is a great idea, and shouldn't be too hard to do as all we'd need to do is append the array. I think we just add it when we need it.